Module scripts should always decode using UTF-8

cdumez opened the pull request:
#### c40873d23eeadf0fe542b79eca7652f9f19689f5
<pre>
Module scripts should always decode using UTF-8
<a href="https://bugs.webkit.org/show_bug.cgi?id=203481">https://bugs.webkit.org/show_bug.cgi?id=203481</a>
rdar://98222621

Reviewed by Youenn Fablet.

By default, CachedScript tries to use the best encoding when decoding script
data. However, per the specification [1], we should always decode as UTF-8 when
it is used as a module (unlike in the classic script case).

This patch adds a ShouldDecodeAsUTF8Only to CachedScript::script() and
scriptHash() and updates CachedScriptSourceProvider pass
ShouldDecodeAsUTF8Only::Yes when used for a module script.

I believe that in theory, the same CachedScript could be used both for a module
script and a classic script, which is why I implemented it this way.

[1] <a href="https://html.spec.whatwg.org/multipage/webappapis.html#fetch-a-single-module-script">https://html.spec.whatwg.org/multipage/webappapis.html#fetch-a-single-module-script</a> (step 12.2)

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/module/charset-01-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/module/charset-02-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/module/charset-03-expected.txt:
Rebaseline tests that are now passing.

* Source/WebCore/bindings/js/CachedScriptSourceProvider.h:
(WebCore::CachedScriptSourceProvider::hash const):
(WebCore::CachedScriptSourceProvider::source const):
(): Deleted.
* Source/WebCore/loader/cache/CachedScript.cpp:
(WebCore::CachedScript::script):
(WebCore::CachedScript::scriptHash):
* Source/WebCore/loader/cache/CachedScript.h:

Canonical link: <a href="https://commits.webkit.org/259251@main">https://commits.webkit.org/259251@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1cc6b7994b3f64dd360c37fe905c21e0829a5442

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104409 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13487 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37318 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113625 "Hash 1cc6b799 for PR 8981 does not build (failure)") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/173918 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108332 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14589 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4404 "Built successfully") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96635 "Hash 1cc6b799 for PR 8981 does not build (failure)") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112663 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110177 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/11258 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94310 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/36/builds/96635 "Hash 1cc6b799 for PR 8981 does not build (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93109 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25921 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/36/builds/96635 "Hash 1cc6b799 for PR 8981 does not build (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6848 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27278 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6977 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/3827 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13004 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46836 "Passed tests") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/6382 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8771 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/bindings/js/CachedScriptSourceProvider.h:
```diff
@@ -57,4 +57,30 @@ class CachedScriptSourceProvider : public JSC::SourceProvider, public CachedReso
+inline unsigned CachedScriptSourceProvider::hash() const
+{
+    switch (sourceType()) {
+    case JSC::SourceProviderSourceType::Module:
```
Seems a bit weird, I would go with a full switch or a if(sourceType() == Module)

cdumez replied:
Ok, will update.

youennf commented on Source/WebCore/bindings/js/CachedScriptSourceProvider.h:
```diff
@@ -57,4 +57,30 @@ class CachedScriptSourceProvider : public JSC::SourceProvider, public CachedReso
+inline StringView CachedScriptSourceProvider::source() const
+{
+    switch (sourceType()) {
+    case JSC::SourceProviderSourceType::Module:
```
Ditto.

youennf reviewed (approved):
Could you add a test where the first decoding fails (say module with UTF-16 data) and we reload the same resource as a classic script?
Either we should not reuse the initial CachedScript or the CachedScript that failed decoding should still have a full m_data so that redecoding succeeds.

cdumez committed:
Module scripts should always decode using UTF-8
https://bugs.webkit.org/show_bug.cgi?id=203481
rdar://98222621

Reviewed by Youenn Fablet.

By default, CachedScript tries to use the best encoding when decoding script
data. However, per the specification [1], we should always decode as UTF-8 when
it is used as a module (unlike in the classic script case).

This patch adds a ShouldDecodeAsUTF8Only to CachedScript::script() and
scriptHash() and updates CachedScriptSourceProvider pass
ShouldDecodeAsUTF8Only::Yes when used for a module script.

I believe that in theory, the same CachedScript could be used both for a module
script and a classic script, which is why I implemented it this way.

[1] https://html.spec.whatwg.org/multipage/webappapis.html#fetch-a-single-module-script (step 12.2)

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/module/charset-01-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/module/charset-02-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/module/charset-03-expected.txt:
Rebaseline tests that are now passing.

* Source/WebCore/bindings/js/CachedScriptSourceProvider.h:
(WebCore::CachedScriptSourceProvider::hash const):
(WebCore::CachedScriptSourceProvider::source const):
(): Deleted.
* Source/WebCore/loader/cache/CachedScript.cpp:
(WebCore::CachedScript::script):
(WebCore::CachedScript::scriptHash):
* Source/WebCore/loader/cache/CachedScript.h:

Canonical link: https://commits.webkit.org/259251@main
